pyro/contrib/funsor/infer/traceenum_elbo.py

```diff
@@ -0,0 +1,82 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import funsor
+
+from pyro.distributions.util import copy_docs_from
+from pyro.infer import ELBO
+from pyro.infer import TraceEnum_ELBO as OrigTraceEnum_ELBO
+from pyro.poutine.util import prune_subsample_sites
+
+from pyro.contrib.funsor import to_data, to_funsor
+from pyro.contrib.funsor.handlers import enum, replay, trace
+
+funsor.set_backend("torch")
+
+
+def expectation(log_probs, costs, sum_vars, prod_vars):
+    result = to_funsor(0, output=funsor.reals())
+    for cost in costs:
+        log_prob = funsor.sum_product.sum_product(
+            funsor.ops.logaddexp, funsor.ops.add, log_probs,
+            plates=prod_vars, eliminate=(prod_vars | sum_vars) - frozenset(cost.inputs)
+        )
+        term = funsor.Integrate(log_prob, cost, sum_vars & frozenset(cost.inputs))
+        term = term.reduce(funsor.ops.add, prod_vars & frozenset(cost.inputs))
+        result += term
+    return result
+
+
+@copy_docs_from(OrigTraceEnum_ELBO)
+class TraceEnum_ELBO(ELBO):
+
+    def _get_trace(self, *args, **kwargs):
+        raise ValueError("shouldn't be here")
+
+    def differentiable_loss(self, model, guide, *args, **kwargs):
```

Comment:
Note the similarity of this function to `funsor.minipyro.elbo`: https://github.com/pyro-ppl/funsor/blob/master/funsor/minipyro.py#L435